orchagent/txmonorch.cpp

```diff
@@ -0,0 +1,313 @@
+                rc = handleThresholdUpdate(key, fvs, true);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown operation type %s when set threshold\n", op.c_str());
+            }
+        }
+
+        if (rc)
+        {
+            SWSS_LOG_ERROR("Handle configuration update failed index %s\n", key.c_str());
+        }
+
+        consumer.m_toSync.erase(it++);
+    }
+}
+
+int TxMonOrch::pollOnePortErrorStatistics(const string &port, TxErrorStatistics &stat)
+{
+    uint64_t txErrStatistics = 0,
+        txErrStatLasttime = tesStatistics(stat),
+        txErrStatThreshold = tesThreshold(stat);
+    int tx_error_state,
+        tx_error_state_lasttime = tesState(stat);
+
+    SWSS_LOG_ENTER();
+
+    {
+        vector<FieldValueTuple> fvs;
+        std::stringstream stream;
+        stream << "oid:0x" << std::hex << tesPortId(stat);
+        std::string keystr( stream.str() );
+        m_countersTable.get(keystr, fvs);
+        for (auto fv: fvs)
+        {
+            if (fvField(fv) == "SAI_PORT_STAT_IF_OUT_ERRORS") {
+                txErrStatistics = stoll(fvValue(fv));
+                break;
+            }
+        }
+
+        SWSS_LOG_INFO("TX_ERR_POLL: got port %s tx_err stati %ld, lasttime %ld threshold %ld\n",
+                      port.c_str(), txErrStatistics, txErrStatLasttime, txErrStatThreshold);
+    }
+
+    if (txErrStatistics - txErrStatLasttime > txErrStatThreshold)
+    {
+        tx_error_state = TXMONORCH_PORT_STATE_ERROR;
+    }
+    else
+    {
+        tx_error_state = TXMONORCH_PORT_STATE_OK;
+    }
+    if (tx_error_state != tx_error_state_lasttime)
+    {
+        tesState(stat) = tx_error_state;
+        //set status in STATE_DB
+        vector<FieldValueTuple> fvs;
+        if (tx_error_state < TXMONORCH_PORT_STATE_MAX)
+            fvs.emplace_back(TXMONORCH_FIELD_STATE_TX_STATE, tx_status_name[tx_error_state]);
+        else
+            fvs.emplace_back(TXMONORCH_FIELD_STATE_TX_STATE, "invalid");
+        m_stateTxErrorTable.set(port, fvs);
+        SWSS_LOG_INFO("TX_ERR_CFG: port %s state changed to %d, push to db\n", port.c_str(), tx_error_state);
+    }
+
+    //refresh the local copy of last time statistics
+    tesStatistics(stat) = txErrStatistics;
+
+    return 0;
+}
+
+void TxMonOrch::pollErrorStatistics()
+{
+    SWSS_LOG_ENTER();
+
+    KeyOpFieldsValuesTuple portEntry;
+
+    for (auto& i : m_PortsTxErrStat)
+    {
+        vector<FieldValueTuple> fields;
+        int rc;
+
+        SWSS_LOG_INFO("TX_ERR_APPL: port %s tx_err_stat %ld, before get\n", i.first.c_str(),
+                        tesStatistics(i.second));
+        rc = pollOnePortErrorStatistics(i.first, i.second);
+        if (rc != 0)
+            SWSS_LOG_ERROR("TX_ERR_APPL: got port %s tx_err_stat failed %d\n", i.first.c_str(), rc);
+        fields.emplace_back(TXMONORCH_FIELD_APPL_STATI, to_string(tesStatistics(i.second)));
+        fields.emplace_back(TXMONORCH_FIELD_APPL_TIMESTAMP, "0");
```

Comment:
not needed

Reply:
I had trouble understanding what this code does.